### PR TITLE
DCS-1749 information about scan numbers for single person

### PR DESCRIPTION
**Pull request by michalnawrockidj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/bodyscan/apiclient/BodyScanApiClient.kt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.beans.factory.annotation.Qualifier
+import org.springframework.stereotype.Component
+import org.springframework.web.reactive.function.client.WebClient
+import uk.gov.justice.digital.hmpps.bodyscan.apiclient.model.PersonalCareCounter
+import uk.gov.justice.digital.hmpps.config.typeReference
+import java.time.LocalDate
+import java.time.format.DateTimeFormatter
+
+@Component
+class BodyScanApiClient(@Qualifier("prisonApiWebClient") private val webClient: WebClient) {
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/bodyscan/apiclient/BodyScanApiClient.kt:12]: we usually name clients over the service they call, so lets call this `BodyScanPrisonApiClient` - this codebase will be the bodyScan service so its a bit weird to call the client bodyscanclient

File: src/main/kotlin/uk/gov/justice/digital/hmpps/bodyscan/model/LimitStatusResponse.kt [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  enum class BodyScanStatus {
+    DO_NOT_SCAN, CLOSE_TO_LIMIT, OK_TO_SCAN
+  }
+
+  @Schema(description = "body scan status", example = "DO_NOT_SCAN")
+  @JsonProperty("bodyScanStatus")
+  fun getBodyScanStatus(): BodyScanStatus {
+    return when {
+      100 > numberOfBodyScans -> BodyScanStatus.OK_TO_SCAN
+      120 < numberOfBodyScans && numberOfBodyScans > 99 -> BodyScanStatus.CLOSE_TO_LIMIT
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/bodyscan/model/LimitStatusResponse.kt:24]: think this should be 116 or more

File: src/main/kotlin/uk/gov/justice/digital/hmpps/bodyscan/resources/LimitStatusResource.kt [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+          )
+        ]
+      ),
+    ]
+  )
+  @GetMapping(path = ["/body-scans/prisoners/{prisonNumber}"])
+  fun getBodyScanForSinglePrisoner(
+    @Schema(description = "Prison Number", example = "G8874VT", required = true)
+    @PathVariable prisonNumber: String
+  ): LimitStatusResponse = limitStatusService.getLimitStatusForYearAndPrisonNumbers(Year.now(), listOf(prisonNumber)).first()
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/bodyscan/resources/LimitStatusResource.kt:67]: throw 404 on not found?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/bodyscan/service/LimitStatusService.kt [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  private val prisonApiClient: BodyScanApiClient
+) {
+  companion object {
+    const val TYPE = "BSCAN"
+  }
+
+  fun getLimitStatusForYearAndPrisonNumbers(year: Year, prisonNumbers: List<String>): List<LimitStatusResponse> {
+    val startDate: LocalDate = LocalDate.ofYearDay(year.value, 1)
+    val endDate: LocalDate = LocalDate.ofYearDay(year.value, startDate.lengthOfYear())
+    val apiMap = prisonApiClient.getPersonalCareNeedsForPrisonNumbers(TYPE, startDate, endDate, prisonNumbers)
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/bodyscan/service/LimitStatusService.kt:20]: meh.. looks like you can do this with adjusters, https://stackoverflow.com/a/33866875

File: src/test/kotlin/uk/gov/justice/digital/hmpps/bodyscan/service/LimitStatusServiceTest.kt [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+  @Test
+  fun `get limit status for year and prison numbers`() {
+    val prisonNumbers = listOf("G8266VG", "G8874VT", "G8874VZ")
+    val year = Year.of(2022)
+    whenever(prisonApiClient.getPersonalCareNeedsForPrisonNumbers(any(), any(), any(), any())).thenReturn(
+      listOf(
+        PersonalCareCounter("G8266VG", 1),
+        PersonalCareCounter("G8874VT", 120),
+        PersonalCareCounter("G8874VZ", 160)
```
Comment (andrewrlee) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/bodyscan/service/LimitStatusServiceTest.kt:27]: it's good practice to test around the boundaries around levels -  if you have a limit of 120 then you would test 119, 120, and 121. 

This highlights fence post errors and inclusivity issues.  

(our limit is 116)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/bodyscan/apiclient/BodyScanApiClient.kt (hunk 1 above)
Comment (michalnawrockidj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/bodyscan/apiclient/BodyScanApiClient.kt:12]: sorted

File: src/main/kotlin/uk/gov/justice/digital/hmpps/bodyscan/model/LimitStatusResponse.kt (hunk 2 above)
Comment (michalnawrockidj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/bodyscan/model/LimitStatusResponse.kt:24]: sorted plus extra tests for edge scenarios

File: src/test/kotlin/uk/gov/justice/digital/hmpps/bodyscan/service/LimitStatusServiceTest.kt (hunk 5 above)
Comment (michalnawrockidj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/bodyscan/service/LimitStatusServiceTest.kt:27]: sorted

File: src/main/kotlin/uk/gov/justice/digital/hmpps/bodyscan/model/LimitStatusResponse.kt [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  enum class BodyScanStatus {
+    DO_NOT_SCAN, CLOSE_TO_LIMIT, OK_TO_SCAN
+  }
+
+  @Schema(description = "body scan status", example = "DO_NOT_SCAN")
+  @JsonProperty("bodyScanStatus")
+  fun getBodyScanStatus(): BodyScanStatus {
+    return when {
+      100 > numberOfBodyScans -> BodyScanStatus.OK_TO_SCAN
+      115 < numberOfBodyScans && numberOfBodyScans > 99 -> BodyScanStatus.CLOSE_TO_LIMIT
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/bodyscan/model/LimitStatusResponse.kt:24]: shouldn't this be less than 116? or <= 115?

**Review by andrewrlee — APPROVED** [review]
LGTM
